pkg/cmd/roachtest/roachtestutil/mixedversion/planner_test.go

```diff
@@ -264,6 +264,94 @@ func TestDeterministicHookSeeds(t *testing.T) {
+// 1. Delays don't change - the delay values in the plan remain unchanged.
+// 2. Idempotent - calling it twice produces the same step count.
+func TestIterateSingleStepsIsPure(t *testing.T) {
+	defer resetMutators()()
+
+	// Create a test with concurrent hooks to ensure we have concurrentRunStep
+	// instances with delays.
+	mvt := newTest(NumUpgrades(1), EnabledDeploymentModes(SystemOnlyDeployment))
+	mvt.InMixedVersion("hook 1", dummyHook)
+	mvt.InMixedVersion("hook 2", dummyHook)
+
+	assertValidTest(mvt, t.Fatal)
+	plan, err := mvt.plan()
+	require.NoError(t, err)
+
+	// Capture plan representation before iteration.
+	planBefore := plan.PrettyPrint()
+
+	// First call to iterateSingleSteps.
+	var stepCount1 int
+	plan.iterateSingleSteps(func(ss *singleStep, isConcurrent bool) {
+		stepCount1++
+	})
+	require.Greater(t, stepCount1, 0, "should have iterated over at least one step")
+
+	// Capture plan representation after first iteration.
+	planAfterFirst := plan.PrettyPrint()
+
+	// Second call to iterateSingleSteps.
+	var stepCount2 int
+	plan.iterateSingleSteps(func(ss *singleStep, isConcurrent bool) {
+		stepCount2++
+	})
+
+	// Capture plan representation after second iteration.
+	planAfterSecond := plan.PrettyPrint()
+
+	// Property 1: Delays don't change - plan representation should be identical.
+	require.Equal(t, planBefore, planAfterFirst,
+		"plan should not change after first iterateSingleSteps call")
+	require.Equal(t, planBefore, planAfterSecond,
+		"plan should not change after second iterateSingleSteps call")
+
+	// Property 2: Idempotent - same step count each time.
+	require.Equal(t, stepCount1, stepCount2,
+		"iterateSingleSteps should return same step count on each call")
+}
+
+// TestMapSingleStepsGeneratesDelays verifies that mapSingleSteps generates
+// new delays when inserting steps into concurrent groups.
+func TestMapSingleStepsGeneratesDelays(t *testing.T) {
+	defer resetMutators()()
+
+	// Create a test with concurrent hooks.
+	mvt := newTest(NumUpgrades(1), EnabledDeploymentModes(SystemOnlyDeployment))
+	mvt.InMixedVersion("hook 1", dummyHook)
+	mvt.InMixedVersion("hook 2", dummyHook)
+
+	assertValidTest(mvt, t.Fatal)
+	plan, err := mvt.plan()
+	require.NoError(t, err)
+
+	// Capture plan representation before mutation.
+	planBefore := plan.PrettyPrint()
+
+	// Use mapSingleSteps to insert a new step into every concurrent group.
+	var insertCount int
+	plan.mapSingleSteps(func(ss *singleStep, isConcurrent bool) []testStep {
+		if isConcurrent {
+			insertCount++
+			newStep := newTestStep(func() error { return nil })
+			return []testStep{ss, newStep}
+		}
+		return []testStep{ss}
+	})
+
+	// Capture plan representation after mutation.
+	planAfter := plan.PrettyPrint()
+
+	// Verify we actually inserted some steps.
+	require.Greater(t, insertCount, 0, "should have inserted at least one step")
+
+	// The plan SHOULD have changed because we inserted new steps.
+	require.NotEqual(t, planBefore, planAfter,
```

Comment:
This test doesn't quite assert that we are generating unique delays, only that we inserted a step. When we chatted in our 1:1, we saw that when I commented out the delay generation the test still passed even though the concurrent steps all had the same delay.

As is I'm not sure if this test is adding additional coverage on top of the datadriven tests that check we can insert mutations. One idea I can think of for how to test we are generating _unique_ delays is to mock `randomConcurrencyDelay` and assert it's called the expected amount of times? I'm not sure if that would be enough for it to be a useful test but it should at least fail in the above example.

Reply:
I have updated the test trying to assign delays to newly inserted concurrent steps post mutation and that original delays are preserved for the original steps.